This is a dataset that describes sonar chirp returns bouncing off different services. The 60 input variables are the strength of the returns at different angles. It is a binary classification problem that requires a model to differentiate rocks from metal cylinders.

In [1]:
# Import essential libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
df = pd.read_csv("sonar_dataset.csv", header=None)
df.sample(5)

0       1       2       3       4   ...      56      57      58      59  60
72   0.0208  0.0186  0.0131  0.0211  0.0610  ...  0.0028  0.0019  0.0049  0.0023   R
29   0.0189  0.0308  0.0197  0.0622  0.0080  ...  0.0099  0.0092  0.0052  0.0075   R
59   0.0125  0.0152  0.0218  0.0175  0.0362  ...  0.0017  0.0041  0.0086  0.0058   R
52   0.0087  0.0046  0.0081  0.0230  0.0586  ...  0.0030  0.0064  0.0058  0.0030   R
131  0.1150  0.1163  0.0866  0.0358  0.0232  ...  0.0190  0.0141  0.0068  0.0086   M

[5 rows x 61 columns]

In [4]:
df.shape

(208, 61)

In [5]:
# check for nan values

df.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [6]:
df.columns

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57, 58, 59, 60],
           dtype='int64')

In [7]:
df[60].value_counts() # label is not skewed

M    111
R     97
Name: 60, dtype: int64

In [8]:
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [9]:
# One hot encoding

y = pd.get_dummies(y, drop_first=True)
y.sample(5) # R --> 1 and M --> 0

R
23   1
174  0
8    1
0    1
66   1

In [10]:
y.value_counts()

R
0    111
1     97
dtype: int64

In [11]:
X.head()

0       1       2       3   ...      56      57      58      59
0  0.0200  0.0371  0.0428  0.0207  ...  0.0180  0.0084  0.0090  0.0032
1  0.0453  0.0523  0.0843  0.0689  ...  0.0140  0.0049  0.0052  0.0044
2  0.0262  0.0582  0.1099  0.1083  ...  0.0316  0.0164  0.0095  0.0078
3  0.0100  0.0171  0.0623  0.0205  ...  0.0050  0.0044  0.0040  0.0117
4  0.0762  0.0666  0.0481  0.0394  ...  0.0072  0.0048  0.0107  0.0094

[5 rows x 60 columns]

In [12]:
# Train test & split

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [13]:
X_train.head()

0       1       2       3   ...      56      57      58      59
67   0.0368  0.0403  0.0317  0.0293  ...  0.0135  0.0067  0.0078  0.0068
14   0.0124  0.0433  0.0604  0.0449  ...  0.0114  0.0196  0.0147  0.0062
164  0.0163  0.0198  0.0202  0.0386  ...  0.0084  0.0108  0.0083  0.0033
179  0.0394  0.0420  0.0446  0.0551  ...  0.0101  0.0068  0.0053  0.0087
19   0.0126  0.0149  0.0641  0.1732  ...  0.0181  0.0094  0.0116  0.0063

[5 rows x 60 columns]

Using Deep Learning Model

In [14]:
# Model without Dropout Layer

import tensorflow as tf
from tensorflow import keras

In [15]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'), # Input layer
    keras.layers.Dense(15, activation='relu'), # Hidden layer
    keras.layers.Dense(1, activation='sigmoid') # Output layer
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 2ms/step - loss: 0.6920 - accuracy: 0.5357
Epoch 2/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6603 - accuracy: 0.6903
Epoch 3/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6302 - accuracy: 0.7073
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5893 - accuracy: 0.7257
Epoch 5/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5717 - accuracy: 0.7870
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5023 - accuracy: 0.8407
Epoch 7/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4940 - accuracy: 0.7897
Epoch 8/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4315 - accuracy: 0.8002
Epoch 9/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4457 - accuracy: 0.8032
Epoch 10/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4654 - accuracy: 0.7976
Epoch 11/

In [17]:
# Teat accuracy

model.evaluate(X_test, y_test)

2/2 [==============================] - 0s 5ms/step - loss: 0.8696 - accuracy: 0.8077


[0.8696040511131287, 0.807692289352417]

Training Accuracy >> Test Accuracy

In [18]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[1.0796332e-09 9.3389022e-01 2.9037124e-01 7.1273230e-06 9.9999964e-01
 9.9983370e-01 9.6966094e-01 9.9999982e-01 1.0313758e-05 1.0000000e+00]
[0. 1. 0. 0. 1. 1. 1. 1. 0. 1.]


In [19]:
y_test[:10]

R
186  0
155  0
165  0
200  0
58   1
34   1
151  0
18   1
202  0
62   1

In [20]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.79      0.85      0.82        27
           1       0.83      0.76      0.79        25

    accuracy                           0.81        52
   macro avg       0.81      0.81      0.81        52
weighted avg       0.81      0.81      0.81        52



In [24]:
# Model with Dropout Layer

modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.1),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.1),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.1),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 2ms/step - loss: 0.7121 - accuracy: 0.5523
Epoch 2/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6563 - accuracy: 0.6031
Epoch 3/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6370 - accuracy: 0.6380
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5970 - accuracy: 0.7111
Epoch 5/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6326 - accuracy: 0.6377
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5775 - accuracy: 0.6862
Epoch 7/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5733 - accuracy: 0.7131
Epoch 8/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5674 - accuracy: 0.7244
Epoch 9/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5093 - accuracy: 0.8071
Epoch 10/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4718 - accuracy: 0.7839
Epoch 11/

In [25]:
modeld.evaluate(X_test, y_test)

2/2 [==============================] - 0s 6ms/step - loss: 0.8487 - accuracy: 0.8269


[0.8487164378166199, 0.8269230723381042]

Training Accuracy is still good but Test Accuracy Improved

In [26]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[1.2603817e-08 4.6542588e-01 1.4110011e-01 1.7134498e-05 9.9991477e-01
 9.9110544e-01 5.0207019e-02 9.9978685e-01 4.5014028e-07 9.9996042e-01]
[0. 0. 0. 0. 1. 1. 0. 1. 0. 1.]


In [27]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.75      1.00      0.86        27
           1       1.00      0.64      0.78        25

    accuracy                           0.83        52
   macro avg       0.88      0.82      0.82        52
weighted avg       0.87      0.83      0.82        52



You can see that by using dropout layer test accuracy increased from 0.81 to 0.83